homeassistant/components/slack/notify.py

```diff
@@ -14,158 +18,226 @@
+    """Define the Slack notification logic."""
+
+    def __init__(self, hass, client, session, default_channel, username, icon):
+        """Initialize."""
+        self._client = client
         self._default_channel = default_channel
-        self._api_token = api_token
-        self._username = username
+        self._hass = hass
         self._icon = icon
+        self._session = session
+        self._username = username
+
         if self._username or self._icon:
             self._as_user = False
         else:
             self._as_user = True
 
-        self.is_allowed_path = is_allowed_path
-        self.slack = Slacker(self._api_token)
-        self.slack.auth.test()
+    async def _async_send_local_file_message(self, path, targets, message, title):
+        """Upload a local file (with message) to Slack."""
+        if not self._hass.config.is_allowed_path(path):
+            _LOGGER.error("Path does not exist or is not allowed: %s", path)
+            return
 
-    def send_message(self, message="", **kwargs):
-        """Send a message to a user."""
+        filename = _async_get_filename_from_url(path)
 
-        if kwargs.get(ATTR_TARGET) is None:
-            targets = [self._default_channel]
-        else:
-            targets = kwargs.get(ATTR_TARGET)
+        try:
+            await self._client.files_upload(
+                channels=",".join(targets),
+                file=path,
+                filename=filename,
+                initial_comment=message,
+                title=title or filename,
+            )
+        except SlackApiError as err:
+            _LOGGER.error("Error while uploading file-based message: %s", err)
+
+    async def _async_send_regular_message(
+        self, targets, message, title, attachments, blocks
+    ):
+        """Send a text-only message."""
+        tasks = {
+            target: self._client.chat_postMessage(
+                channel=target,
+                text=message,
+                as_user=self._as_user,
+                attachments=attachments,
+                blocks=blocks,
+                icon_emoji=self._icon,
+                link_names=True,
+            )
+            for target in targets
+        }
+
+        results = await asyncio.gather(*tasks.values(), return_exceptions=True)
+        for target, result in zip(tasks, results):
+            if isinstance(result, SlackApiError):
+                _LOGGER.error(
+                    "There was a Slack API error while sending a message to %s: %s",
+                    target,
+                    result,
+                )
+
+    async def _async_send_remote_file_message(
+        self, url, targets, message, title, username, password, auth_type
+    ):
+        """Upload a remote file (with message) to Slack."""
+        try:
+            file_as_bytes = await self._hass.async_add_executor_job(
+                _get_remote_file_contents, url, username, password, auth_type
```

Comment:
This feature needs to be removed. It's dangerous. It means that we're allowing via a service call any file to be fetched in the name of Home Assistant (which can have elevated privileges because it's making internal requests).

Reply:
We've had other integrations remove it too. 

An architecture issue can be opened to see if we can add a url whitelist or something like that.

Reply:
Roger that. How do you feel about my removing it from this PR as a breaking change and doing the architectural issue in parallel?

Reply:
Is this true for local files in whitelisted dirs, as well? Or only for remote files?

Reply:
af45468 removes the remote option; let me know if I'm okay to keep the local option.

Reply:
local files in whitelisted dirs are ok. URLs have no whitelist and that is dangerous.